File: sigma_c/bindings/SigmaC.m
(* Sigma-C Mathematica Package *)
(* ============================= *)
(* Copyright (c) 2025 ForgottenForge.xyz *)

BeginPackage["SigmaC`"]

CriticalPoint::usage = "CriticalPoint[data] computes the critical point from {epsilon, observable} data."
ComputeSigmaC::usage = "ComputeSigmaC[epsilon, observable] computes criticality metrics."
ListPlotCritical::usage = "ListPlotCritical[data] plots data with critical point highlighted."
StreamingSigmaC::usage = "StreamingSigmaC[windowSize] creates a streaming calculator."

Begin["`Private`"]

(* Core computation *)
ComputeSigmaC[epsilon_List, observable_List] := Module[
    {chi, peakIdx, sigmaC, chiMax, kappa},
    
    (* Compute susceptibility *)
    chi = Differences[observable] / Differences[epsilon];
    
    (* Find peak *)
    peakIdx = First[Ordering[Abs[chi], -1]];
    sigmaC = epsilon[[peakIdx]];
    chiMax = Abs[chi[[peakIdx]]];
    
    (* Compute kappa *)
    kappa = If[Length[chi] > 2,
        Abs[Differences[chi][[peakIdx]]] / chiMax,
        1.0
    ];
    
    <|
        "SigmaC" -> sigmaC,
        "Kappa" -> kappa,
        "ChiMax" -> chiMax,
        "PeakLocation" -> sigmaC
    |>
]

(* Convenience function *)
CriticalPoint[data_List] := Module[
    {epsilon, observable},
    epsilon = data[[All, 1]];
    observable = data[[All, 2]];
    ComputeSigmaC[epsilon, observable]
]

(* Visualization *)
ListPlotCritical[data_List, opts___] := Module[
    {result, sigmaC, plot},
    
    result = CriticalPoint[data];
    sigmaC = result["SigmaC"];
    
    plot = ListPlot[data,
        opts,
        Epilog -> {
            Red, PointSize[0.02],
            Point[{sigmaC, Interpolation[data][sigmaC]}],
            Text[Style[StringForm["σ_c = ``", NumberForm[sigmaC, {4, 3}]], 14],
                {sigmaC, Interpolation[data][sigmaC]}, {0, -2}]
        },
        PlotLabel -> "Criticality Analysis",
        AxesLabel -> {"ε", "Observable"}
    ];
    
    plot
]

(* Streaming calculator *)
StreamingSigmaC[windowSize_: 100] := Module[
    {buffer = {}, currentSigmaC = 0.0},
    
    Function[{epsilon, observable},
        AppendTo[buffer, {epsilon, observable}];
        
        (* Keep only recent window *)
        If[Length[buffer] > windowSize,
            buffer = Take[buffer, -windowSize]
        ];
        
        (* Recompute *)
        If[Length[buffer] >= 10,
            result = CriticalPoint[buffer];
            currentSigmaC = result["SigmaC"]
        ];
        
        currentSigmaC
    ]
]

End[]
EndPackage[]

(* Example usage:
   data = Table[{x, Sin[10 x]}, {x, 0, 0.5, 0.01}];
   result = CriticalPoint[data];
   ListPlotCritical[data]
*)

File: sigma_c/bindings/SigmaC.jl
# Sigma-C Julia Package
# ======================
# Copyright (c) 2025 ForgottenForge.xyz

"""
    SigmaC

Julia package for criticality analysis.

# Examples
```julia
using SigmaC

# Compute criticality
epsilon = range(0, 0.5, length=100)
observable = sin.(epsilon * 10)
result = compute_sigma_c(epsilon, observable)

println("σ_c = ", result.sigma_c)
```
"""
module SigmaC

export compute_sigma_c, StreamingSigmaC, ObservableDiscovery

using Statistics
using FFTW

"""
    SigmaCResult

Result structure for criticality analysis.
"""
struct SigmaCResult
    sigma_c::Float64
    kappa::Float64
    chi_max::Float64
    peak_location::Float64
end

"""
    compute_sigma_c(epsilon::AbstractVector, observable::AbstractVector)

Compute critical susceptibility from data.

# Arguments
- `epsilon`: Control parameter values
- `observable`: Observable values

# Returns
- `SigmaCResult` with sigma_c, kappa, and other metrics
"""
function compute_sigma_c(epsilon::AbstractVector, observable::AbstractVector)
    # Compute susceptibility (derivative)
    chi = diff(observable) ./ diff(epsilon)
    
    # Find peak
    peak_idx = argmax(abs.(chi))
    sigma_c = epsilon[peak_idx]
    chi_max = abs(chi[peak_idx])
    
    # Compute peak sharpness (kappa)
    # Using second derivative
    if length(chi) > 2
        d2_chi = diff(chi)
        kappa = abs(d2_chi[peak_idx]) / chi_max
    else
        kappa = 1.0
    end
    
    return SigmaCResult(sigma_c, kappa, chi_max, sigma_c)
end

"""
    StreamingSigmaC

Streaming criticality calculator with O(1) updates.
"""
mutable struct StreamingSigmaC
    window_size::Int
    epsilon_buffer::Vector{Float64}
    observable_buffer::Vector{Float64}
    current_sigma_c::Float64
    
    function StreamingSigmaC(window_size::Int=100)
        new(window_size, Float64[], Float64[], 0.0)
    end
end

"""
    update!(stream::StreamingSigmaC, epsilon::Float64, observable::Float64)

Update streaming calculator with new data point.
"""
function update!(stream::StreamingSigmaC, epsilon::Float64, observable::Float64)
    push!(stream.epsilon_buffer, epsilon)
    push!(stream.observable_buffer, observable)
    
    # Keep only recent window
    if length(stream.epsilon_buffer) > stream.window_size
        popfirst!(stream.epsilon_buffer)
        popfirst!(stream.observable_buffer)
    end
    
    # Recompute sigma_c
    if length(stream.epsilon_buffer) >= 10
        result = compute_sigma_c(stream.epsilon_buffer, stream.observable_buffer)
        stream.current_sigma_c = result.sigma_c
    end
    
    return stream.current_sigma_c
end

"""
    ObservableDiscovery

Automatic observable discovery using gradient/entropy methods.
"""
struct ObservableDiscovery
    method::Symbol  # :gradient, :entropy, or :pca
end

"""
    find_optimal(discovery::ObservableDiscovery, data::Matrix)

Find optimal observable from candidate features.

# Arguments
- `data`: Matrix where each column is a candidate observable

# Returns
- Index of best observable
"""
function find_optimal(discovery::ObservableDiscovery, data::Matrix)
    n_features = size(data, 2)
    scores = zeros(n_features)
    
    for i in 1:n_features
        if discovery.method == :gradient
            # Maximize gradient norm
            scores[i] = sum(abs.(diff(data[:, i])))
        elseif discovery.method == :entropy
            # Minimize entropy
            hist = fit(Histogram, data[:, i], nbins=20)
            p = hist.weights ./ sum(hist.weights)
            p = p[p .> 0]  # Remove zeros
            scores[i] = -sum(p .* log.(p))
        end
    end
    
    return argmax(scores)
end

end # module
